python/cudf/cudf/core/frame.py

```diff
@@ -1819,6 +1827,271 @@ def sample(
+                    cudf_dictionaries_frame._data[name],
+                    codes,
+                    mask=codes.base_mask,
+                    size=codes.size,
+                    ordered=dict_ordered[name],
+                )
+
+        # Handle non-dict arrays
+        cudf_non_category_frames = (
+            cudf.DataFrame()
+            if data.num_columns == 0
+            else cudf.DataFrame._from_table(
+                libcudf.interop.from_arrow(data, data.column_names)
+            )
+        )
+
+        if (
+            cudf_non_category_frames.shape[1] > 0
+            and cudf_category_frames.shape[1] > 0
+        ):
+            result = cudf.concat(
+                [cudf_non_category_frames, cudf_category_frames], axis=1
+            )
+        elif cudf_non_category_frames.shape[1] > 0:
+            result = cudf_non_category_frames
+        else:
+            result = cudf_category_frames
+
+        if dtypes:
+            for name in result._data.names:
+                dtype = dtypes[name]
+                if dtype == "categorical":
+                    dtype = "category"
+                elif dtype == "date" or dtype == "datetime":
+                    dtype = "datetime64[ms]"
+                elif dtype == "empty":
+                    dtype = "object"
+
+                result._data[name] = result._data[name].astype(dtype)
+
+        return result[column_names]
+
+    @annotate("TO_ARROW", color="orange", domain="cudf_python")
+    def to_arrow(self):
+        """
+        Convert to arrow Table
+
+        Examples
+        --------
+        >>> import cudf
+        >>> df = cudf.DataFrame(
+        ...     {"a":[1, 2, 3], "b":[4, 5, 6]}, index=[1, 2, 3])
+        >>> df.to_arrow()
+        pyarrow.Table
+        a: int64
+        b: int64
+        index: int64
+        """
+
+        data = self.copy(deep=False)
+
+        codes = {}
+        # saving the name as they might get changed from int to str
+        codes_keys = []
+        categories = {}
+        # saving the name as they might get changed from int to str
+        names = self._data.names
+        null_arrays_names = []
+
+        for name in self._data.names:
+            col = self._data[name]
+            if isinstance(col, cudf.core.column.CategoricalColumn,):
+                # arrow doesn't support unsigned codes
+                signed_type = (
+                    min_signed_type(col.codes.max())
+                    if col.codes.size > 0
+                    else np.int8
+                )
+                codes[str(name)] = col.codes.astype(signed_type)
+                categories[str(name)] = col.categories
+                codes_keys.append(name)
+            elif isinstance(
+                col, cudf.core.column.StringColumn
+            ) and col.null_count == len(col):
+                null_arrays_names.append(name)
+
+        data = cudf.DataFrame(data._data)
+        data = data.drop(codes_keys + null_arrays_names, inplace=True)
```

Comment:
It would be good to avoid using a `cudf.DataFrame` internally if possible here. This is effectively creating a cyclic dependency.

Reply:
Was this not possible to avoid?